Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -26,51 +26,368 @@
+            
+            searchRange.start = newStart.value();
+            matchRange = makeSimpleRange(prefixMatch->end, searchRange.end);
+            advanceRangeStartToNextNonWhitespace(matchRange.value());
+            
+            if (matchRange->collapsed())
+                return std::nullopt;
+            
+            ASSERT(matchRange->start.container->isTextNode(), "MatchRange start is not a Text Node");
+            
+            potentialMatch = rangeOfStringInRange(parsedTextDirective.textStart, matchRange.value(), false, mustEndAtWordBoundary);
+
+            if (!potentialMatch)
+                return std::nullopt;
+            if (potentialMatch->start != matchRange->start) {
+                continue;
+            }
+        }
+        else
+        {
+            potentialMatch = rangeOfStringInRange(parsedTextDirective.textStart, searchRange, true, mustEndAtWordBoundary);
+            if (!potentialMatch)
+                return std::nullopt;
+            
+            auto newStart = makeBoundaryPoint(makeContainerOffsetPosition(potentialMatch->start).next());
+            if (!newStart)
+                return std::nullopt;
+            searchRange.start = newStart.value();
+        }
+        auto rangeEndSearchRange = makeSimpleRange(potentialMatch->end, searchRange.end);
+        while (!rangeEndSearchRange.collapsed())
+        {
+            if (!parsedTextDirective.textEnd.isNull())
+            {
+                mustEndAtWordBoundary = !parsedTextDirective.suffix;
+                auto textEndMatch = rangeOfStringInRange(parsedTextDirective.textEnd, rangeEndSearchRange, true, mustEndAtWordBoundary);
+                if (!textEndMatch)
+                    return std::nullopt;
+                
+                potentialMatch->end = textEndMatch->end;
+            }
+            // TODO: Assert: potentialMatch represents a range exactly containing an instance of matching text.
+            RELEASE_ASSERT_WITH_MESSAGE(potentialMatch && !potentialMatch->collapsed(), "Scroll To Text Fragment: potentialMatch cannot be null or collapsed");
+            
+            if (!parsedTextDirective.suffix)
+                return potentialMatch;
+            
+            auto suffixRange = makeSimpleRange(potentialMatch->end, searchRange.end);
+            advanceRangeStartToNextNonWhitespace(suffixRange);
+            auto suffixMatch = rangeOfStringInRange(parsedTextDirective.suffix, suffixRange, false, true);
+            
+            if (!suffixMatch)
+                return std::nullopt;
+            
+            if (suffixMatch->start == suffixRange.start)
+                return potentialMatch;
+            
+            if (parsedTextDirective.textEnd.isNull())
+                break;
+            
+            rangeEndSearchRange.start = potentialMatch->end;
+        }
+        
+        if (rangeEndSearchRange.collapsed()) {
+            ASSERT(!parsedTextDirective.textEnd.isNull());
```

Comment:
Do you want this assertion to run if `rangeEndSearchRange.collapsed()` is true before we start looping?

Reply:
This is matching the spec:

If rangeEndSearchRange is collapsed then:
Assert: parsedValues’s textEnd item is non-null
Return null

Reply:
Right but what if rangeEndSearchRange is collapsed before we begin looping? This code won't run.

Reply:
My mistake, we are outside of the `while` loop not inside.